cirq-core/cirq/ops/control_values.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+from typing import Union, Tuple, List, TYPE_CHECKING, Any, Dict, Generator, Optional, cast, Iterator
+from dataclasses import dataclass
+
+import itertools
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class AbstractControlValues(abc.ABC):
+    """AbstractControlValues is an abstract immutable data class.
+
+    AbstractControlValues defines an API for control values and implements
+    functions common to all implementations (e.g.  comparison).
+    """
+
+    @abc.abstractmethod
+    def __and__(self, other: 'AbstractControlValues') -> 'AbstractControlValues':
+        """Sets self to be the cartesian product of all combinations in self x other.
+
+        Args:
+          other: An object that implements AbstractControlValues.
+
+        Returns:
+          An object that represents the cartesian product of the two inputs.
+        """
+
+    @abc.abstractmethod
+    def _expand(self) -> Iterator[Tuple[Any, ...]]:
+        """Returns the control values tracked by the object."""
```

Comment:
Please explain more what the "expanded" representation means / contains. Does it depend upon what internal representation the subclasses use? Does every tuple correspond to a "one value per qubit" representation?